pkg/types/types.go

```diff
@@ -0,0 +1,377 @@
+			workerMatrix.Matrix = append(workerMatrix.Matrix, entry)
+		}
+	}
+
+	return masterMatrix, workerMatrix
+}
+
+func (m *ComMatrix) writeMatrixToFile(utilsHelpers utils.UtilsInterface, fileName, format string, destDir string) error {
+	res, err := m.print(format)
+	if err != nil {
+		return err
+	}
+
+	comMatrixFileName := filepath.Join(destDir, fmt.Sprintf("%s.%s", fileName, format))
+	return utilsHelpers.WriteFile(comMatrixFileName, res)
+}
+
+// Diff returns the diff ComMatrix.
+func (m *ComMatrix) Diff(other ComMatrix) ComMatrix {
+	diff := []ComDetails{}
+	for _, cd1 := range m.Matrix {
+		found := false
+		for _, cd2 := range other.Matrix {
+			if cd1.Equals(cd2) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			diff = append(diff, cd1)
+		}
+	}
+
+	return ComMatrix{Matrix: diff}
+}
+
+func (m *ComMatrix) Contains(cd ComDetails) bool {
+	for _, cd1 := range m.Matrix {
+		if cd1.Equals(cd) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (m *ComMatrix) ToNFTables() ([]byte, error) {
+	var tcpPorts []string
+	var udpPorts []string
+	for _, line := range m.Matrix {
+		if line.Protocol == "TCP" {
+			tcpPorts = append(tcpPorts, fmt.Sprint(line.Port))
+		} else if line.Protocol == "UDP" {
+			udpPorts = append(udpPorts, fmt.Sprint(line.Port))
+		}
+	}
+
+	tcpPortsStr := strings.Join(tcpPorts, ", ")
+	udpPortsStr := strings.Join(udpPorts, ", ")
+
+	result := fmt.Sprintf(`#!/usr/sbin/nft -f
+
+	table inet openshift_filter {
+		chain OPENSHIFT {
+			type filter hook input priority 1; policy accept;
+
+			# Allow loopback traffic
+			iif lo accept
+	
+			# Allow established and related traffic
+			ct state established,related accept
+	
+			# Allow ICMP on ipv4
+			ip protocol icmp accept
+			# Allow ICMP on ipv6
+			ip6 nexthdr ipv6-icmp accept
+
+			# Allow specific TCP and UDP ports
+			tcp dport  { %s } accept
+			udp dport { %s } accept
+
+			# Logging and default drop
+			log prefix "firewall " drop
+		}
+	}`, tcpPortsStr, udpPortsStr)
+
+	return []byte(result), nil
+}
+
+func (cd ComDetails) String() string {
+	return fmt.Sprintf("%s,%s,%d,%s,%s,%s,%s,%s,%v", cd.Direction, cd.Protocol, cd.Port, cd.Namespace, cd.Service, cd.Pod, cd.Container, cd.NodeRole, cd.Optional)
+}
+
+func (cd ComDetails) Equals(other ComDetails) bool {
+	strComDetail1 := fmt.Sprintf("%s-%d-%s", cd.NodeRole, cd.Port, cd.Protocol)
+	strComDetail2 := fmt.Sprintf("%s-%d-%s", other.NodeRole, other.Port, other.Protocol)
+
+	return strComDetail1 == strComDetail2
+}
+
+func getComMatrixHeadersByFormat(format string) (string, error) {
+	typ := reflect.TypeOf(ComDetails{})
+
+	var tagsList []string
+	for i := 0; i < typ.NumField(); i++ {
+		field := typ.Field(i)
+		tag := field.Tag.Get(format)
+		if tag == "" {
+			return "", fmt.Errorf("field %v has no tag of format %s", field, format)
+		}
+		tagsList = append(tagsList, tag)
+	}
+
+	return strings.Join(tagsList, ","), nil
+}
+
+func CleanComDetails(outPuts []ComDetails) []ComDetails {
```

Comment:
Suggestion: Changing this func to a method of  `ComMatrix`, that applies the `ComDetails` list changes on the matrix instead of returning it. 
What do you think?

Reply:
yah i agree with shir since the [] ComDetails is a ComMatrix

Reply:
sounds good :)